### ISSUE TITLE
Complex max/min fail on shared gpu device arrays

### ISSUE THREAD

**Issue by inailuig** [google/jax#16551]
### Description

Use case: I need to use `jax.scipy.special.logsumexp` on a complex array, shared across multiple gpu's.
However this fails with an error message equal to the one below. 

I have been able to trace it down to the max inside of the aforementioned function erroring on complex arrays (distributed over multiple gpu's), as can be seen from the following reproducer:

```python
import jax
import jax.numpy as jnp

x = jnp.ones(128, dtype=jnp.complex64)
sharding = jax.sharding.PositionalSharding(jax.devices())
x = jax.device_put(x, sharding)
jax.debug.visualize_array_sharding(x)
jax.jit(jnp.max)(x)
#jax.jit(jnp.min)(x)

```
output:
```
  GPU 0    GPU 1  
                  
2023-06-24 23:32:07.052799: E external/xla/xla/status_macros.cc:54] INTERNAL: RET_CHECK failure (external/xla/xla/service/gpu/nccl_all_reduce_thunk.cc:110) mlir::isa<mlir::mhlo::ConvertOp>(result_op) 
*** Begin stack trace ***
	[...]
*** End stack trace ***

2023-06-24 23:32:07.053275: E external/xla/xla/status_macros.cc:54] INTERNAL: RET_CHECK failure (external/xla/xla/service/gpu/nccl_all_reduce_thunk.cc:110) mlir::isa<mlir::mhlo::ConvertOp>(result_op) 
*** Begin stack trace ***
	[...]
*** End stack trace ***

2023-06-24 23:32:07.053329: F external/xla/xla/service/gpu/nccl_all_reduce_thunk.cc:152] Check failed: reduction_kind.has_value() 
Aborted (core dumped)```

[trace.txt](https://github.com/google/jax/files/11858302/trace.txt)

### What jax/jaxlib version are you using?

jax-0.4.13 jaxlib-0.4.13+cuda12.cudnn89

### Which accelerator(s) are you using?

GPU

### Additional system info

libcudnn8 8.9.2.26-1+cuda12.1, libnccl2 2.18.3-1+cuda12.1

### NVIDIA GPU info

```   
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 525.105.17   Driver Version: 525.105.17   CUDA Version: 12.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA A100-SXM...  On   | 00000000:4C:00.0 Off |                    0 |
| N/A   26C    P0    61W / 400W |      2MiB / 81920MiB |      0%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
|   1  NVIDIA A100-SXM...  On   | 00000000:88:00.0 Off |                    0 |
| N/A   25C    P0    65W / 400W |      2MiB / 81920MiB |      0%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
```

**Comment by yashk2810** [comment]
This looks like a XLA bug to me.

**Comment by inailuig** [comment]
I understand that it will not be straightforward to implement this, given these operations are not supported by nccl, but would it be possible to at least get a reasonable error message instead of the hard crash?

**Comment by inailuig** [comment]
For now I have been using a workaround which roughly looks like this, using allgather and then reducing locally
```python
import jax
import jax.numpy as jnp
from jax.sharding import Mesh, PartitionSpec as P
from jax.experimental.shard_map import shard_map
from functools import partial

def collective_max(x, axis=None, keepdims=False):
    mesh = Mesh(jax.devices(), axis_names=('i'))

    @partial(shard_map, mesh=mesh, in_specs=(P('i')), out_specs=P(None))
    def _collective_max(x):
        tmp1 = jnp.max(x, axis=axis, keepdims=keepdims)
        tmp2 = tmp1[None]
        tmp3 = jax.lax.all_gather(tmp2, 'i', tiled=True)
        return jnp.max(tmp3, axis=0)

    return _collective_max(x)
```

**Comment by hawkinsp** [comment]
Basic question: why are we computing the *maximum* of a list of complex numbers? Maximum in the numpy world is defined as the maximum on lexicographic `(real, imag)` pairs. But that seems nonsensical in the context of `logsumexp`?

Given the point `max` calculation is done for numerical stability, perhaps it should be doing something like taking the number with the maximum absolute value? That would be supported by the existing all-reduce implementation.

**Comment by inailuig** [comment]
> Basic question: why are we computing the _maximum_ of a list of complex numbers? Maximum in the numpy world is defined as the maximum on lexicographic `(real, imag)` pairs. But that seems nonsensical in the context of `logsumexp`?

I agree. If we were to we would just to implement it for consistency.

> Given the `max` calculation is done for numerical stability, perhaps it should be doing something like taking the number with the maximum absolute value? That would be supported by the existing all-reduce implementation.

Looks like this goes in the right direction but does not quite seem like the correct approach. 
Thinking about it, what we should do is to shift by the (negative) maximum value of the real part. This way amplitude after the first exp is <= 1. 
A shift of the imag part (i.e. taking out a global phase) seems irrelevant numerically, at least if implemented in the trivial way with the euler formula, e.g. like in libstdc++: 
https://github.com/llvm/llvm-project/blob/0894c2ee5f602e1766c8d8c5911797bf31cb8c4a/libcxx/include/complex#L1060-L1061

The only disadvantage of only shifting the real part is that we will get different results compared to numpy, as the phase of the sum (imag part of the result) might be shifted by multiples of 2π.

@hawkinsp If you want I can open a PR with this proposal, let me know.
